File: section_05/code_5_5_Frog.swift
import Foundation

// Frog問題 (配る遷移形式)

let INF = Int.max
let h = [2, 9, 4, 5, 1, 6, 10]

// 足場0から足場iまでの最小コストを返す
func walk(_ list: [Int], _ i: Int) -> Int {
    // ベースケース
    if i == 0 {
        return 0
    }

    // 1つ前からの移動コスト
    let c1 = walk(list, i - 1) + abs(list[i] - list[i - 1])

    // 2つ前からの移動コスト
    var c2 = INF
    if i - 2 >= 0 {
        c2 = walk(list, i - 2) + abs(list[i] - list[i - 2])
    }

    return min(c1, c2)
}

// print(h)
print(walk(h, h.count - 1))

File: section_05/code_5_5_Frog.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

#
# Frog問題. 再帰を使って全探索
#

INF = 1 << 60

# 足場0からiまでの最小コスト
def walk(list, i)
  # ベースケース (足場のコストは0)
  return 0 if i == 0

  # 1つ前から来た場合のコスト
  v = walk(list, i - 1) + (list[i] - list[i - 1]).abs

  # 2つ前から来た場合のコスト
  v2  = INF
  if i > 1
    v2 = walk(list, i - 2) + (list[i] - list[i - 2]).abs
  end

  # 小さい方が足場0からの最小コスト
  return [v, v2].min
end

h = [2, 9, 4, 5, 1, 6, 10]
pp walk(h, h.size - 1)  # 8
